includes/class-wc-payments-token-service.php

```diff
@@ -213,6 +205,77 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
 		return $tokens;
 	}
 
+	/**
+	 * Retrieves the payment method types for which tokens should be retrieved.
+	 *
+	 * This function determines the appropriate payment method types based on the provided gateway ID.
+	 * - If a gateway ID is provided, it retrieves the payment methods specific to that gateway to prevent duplication of saved tokens under incorrect payment methods during checkout.
+	 * - If no gateway ID is provided, it retrieves the default payment methods to fetch all saved tokens, e.g., for the Blocks checkout or My Account page.
+	 *
+	 * @param string|null $gateway_id The optional ID of the gateway.
+	 * @return array The list of retrievable payment method types.
+	 */
+	private function get_retrievable_payment_method_types( $gateway_id = null ) {
+		if ( empty( $gateway_id ) ) {
+			return $this->get_all_retrievable_payment_types();
+		} else {
+			return $this->get_gateway_specific_retrievable_payment_types( $gateway_id );
+		}
+	}
+
+	/**
+	 * Returns all the enabled retrievable payment method types.
+	 *
+	 * @return array Enabled retrievable payment method types.
+	 */
+	private function get_all_retrievable_payment_types() {
+		$types = [ Payment_Method::CARD ];
+
+		if ( $this->is_payment_method_enabled( Payment_Method::SEPA ) ) {
+			$types[] = Payment_Method::SEPA;
+		}
+
+		if ( $this->is_payment_method_enabled( Payment_Method::LINK ) ) {
+			$types[] = Payment_Method::LINK;
+		}
+
+		return $types;
+	}
```

Comment:
This is the previous implementation w/o any changes, just that it's being run only when no gateway is provided, meaning that we're retrieving tokens as a single list at once.